New non-working driver: SONY DFS-500 DME Video Mixer (1994)

felipesanches opened the pull request:
And a few new skeleton drivers for some other SONY video equipment:
* U-Matic vo5850PM
* BETACAM-SP UVW-1200
* BETACAM-SP UVW-1600
* BETACAM-SP UVW-1800

Note: This is something I've been working on my free time throughout the last month. I've put a lot of effort on the DFS-500 driver during the past weekends and I believe I reached a milestone in its development that makes it worthy to submit this PR. I will definitely work more on this one, but for now I'd like to submit this and rest a bit before putting more effort into it.

![Screenshot from 2020-10-13 15-17-06](https://user-images.githubusercontent.com/213676/95899633-395b2800-0d67-11eb-99de-f58850ee938f.png)

Osso13 commented:
Leaving the actual PR review to someone better than me, but please consider adding the new driver files to src/mame/mess.flt, so that people that still use the mess TARGET can have them.

cuavas commented on src/mame/drivers/betacam.cpp:
```diff
@@ -0,0 +1,135 @@
+
+
+static INPUT_PORTS_START(betacam)
+  PORT_START("DSW1")
+    PORT_DIPUNUSED_DIPLOC( 0x01, 0x00, "SW1:1" )
+    PORT_DIPNAME( 0x02, 0x00, "RGB Output Sel" ) PORT_DIPLOCATION("SW1:2")
+    PORT_DIPSETTING(    0x00, DEF_STR( No ) )
+    PORT_DIPSETTING(    0x02, DEF_STR( Yes ) )
+    PORT_DIPNAME( 0x04, 0x00, "RGB Input Sel" ) PORT_DIPLOCATION("SW1:3")
+    PORT_DIPSETTING(    0x04, DEF_STR( Yes ) )
+    PORT_DIPSETTING(    0x00, DEF_STR( No ) )
+    PORT_DIPNAME( 0x08, 0x00, "Wide" ) PORT_DIPLOCATION("SW1:4")
+    PORT_DIPSETTING(    0x08, DEF_STR( Yes ) )
+    PORT_DIPSETTING(    0x00, DEF_STR( No ) )
+    PORT_DIPNAME( 0x10, 0x00, "R/P E/F" ) PORT_DIPLOCATION("SW1:5")
+    PORT_DIPSETTING(    0x10, DEF_STR( Yes ) )
+    PORT_DIPSETTING(    0x00, DEF_STR( No ) )
+    PORT_DIPNAME( 0x20, 0x20, "Rec/Player" ) PORT_DIPLOCATION("SW1:6")
+    PORT_DIPSETTING(    0x20, DEF_STR( Yes ) )
+    PORT_DIPSETTING(    0x00, DEF_STR( No ) )
+    PORT_DIPNAME( 0x40, 0x00, "J U/C" ) PORT_DIPLOCATION("SW1:7")
+    PORT_DIPSETTING(    0x40, DEF_STR( Yes ) )
+    PORT_DIPSETTING(    0x00, DEF_STR( No ) )
+    PORT_DIPNAME( 0x80, 0x00, "NTSC / PAL" ) PORT_DIPLOCATION("SW1:8")
+    PORT_DIPSETTING(    0x80, DEF_STR( Yes ) )
+    PORT_DIPSETTING(    0x00, DEF_STR( No ) )
+INPUT_PORTS_END
```
Please fix the indentation here and elsewhere.  Use one tab per level of scope.

cuavas commented on src/mame/drivers/dfs500.cpp:
```diff
@@ -0,0 +1,1036 @@
+    selector buttons VTR A, VTR B, 3 and 4 (to select ammong the 4 video inputs).
+
+****************************************************************************/
+#define VIDEO_WIDTH 768
+#define VIDEO_HEIGHT 256
```
Don’t put anything in a compilation unit besides comments before you include the prefix header.

felipesanches replied:
what is "the prefix header"?

Robbbert replied:
I think he means the include of emu.h

felipesanches replied:
if that's the case, then now we're good :-)

cuavas replied:
emu.h is pre-compiled and supplied as a prefix header in the default configuration.

cuavas commented on src/mame/drivers/dfs500.cpp:
```diff
@@ -0,0 +1,1036 @@
+	void dfs500(machine_config &config);
+
+private:
+	virtual void machine_start() override;
```
Overridden protected members should be protected.

cuavas commented on src/mame/drivers/dfs500.cpp:
```diff
@@ -0,0 +1,1036 @@
+u32 dfs500_state::screen_update(screen_device &screen, bitmap_rgb32 &bitmap, rectangle const &cliprect)
+{
+	bitmap_argb32 *input_bitmap;
+	switch(m_input_sel_A & 3)
+	{
+		case 0: input_bitmap = &m_input1->get_bitmap(); break;
+		case 1: input_bitmap = &m_input2->get_bitmap(); break;
+		case 2: input_bitmap = &m_input3->get_bitmap(); break;
+		case 3: input_bitmap = &m_input4->get_bitmap(); break;
+	}
```
Use a `required_device_array` to avoid the switch statement.

cuavas commented on src/mame/drivers/dfs500.cpp:
```diff
@@ -0,0 +1,1036 @@
+		{
+			for (unsigned screen_x = screen.visible_area().min_x; screen_x <= screen.visible_area().max_x; screen_x++)
+			{
+				bitmap.pix(screen_y, screen_x) = input_bitmap->pix((int)((double)(screen_y % VIDEO_HEIGHT) * stepy),
+				                                                   (int)((double)(screen_x % VIDEO_WIDTH) * stepx));
```
Don’t try aligning stuff like this – it displays incorrectly when people use different tab sizes.  Just use two tabs for line continuations.

felipesanches replied:
The second line contains tabs up until the letter b of `bitmap` of the first one and then spaces for the intended alignment. It will work in all possible text editor setups.

cuavas replied:
You can see that we run an automated whitespace normalisation tool on the source before forking each release.  Indentation outside comments is converted to tabs.  The kind of formatting you’ve used here will not be preserved.

If you need to break up the arguments across lines, indent them two tabs from the level of the call so it’s clear it’s a statement continuation and not a new scope level.  Also, lose some of the superfluous parentheses.
```
				bitmap.pix(screen_y, screen_x) = input_bitmap->pix(
						int(double(screen_y % VIDEO_HEIGHT) * stepy),
						int(double(screen_x % VIDEO_WIDTH) * stepx));
```

felipesanches replied:
done

cuavas commented on src/mame/drivers/dfs500.cpp:
```diff
@@ -0,0 +1,1036 @@
+uint8_t dfs500_state::RB2_r(offs_t offset)
+{
+	uint8_t value = 0;
+	value |= ((ioport("DSW_S1")->read() & 0x0f) << 4); // (Editing Control Unit Select)
+	value |= (ioport("DSW_S2")->read() & 0x0f);        // (Freeze Timing)
```
Don’t do tag lookups on the hot path - they’re expensive, and you get no validation.

felipesanches replied:
done

cuavas commented on src/mame/drivers/dfs500.cpp:
```diff
@@ -0,0 +1,1036 @@
+			// D1 = LD186 or LD168 (?) = DP on pattern_number10
+			// D0 = LD158 = DP on pattern_number1
+			m_dot_points2 = data;
+			output().set_value("trans_rate100", ls247_map[m_trans_rate[1] & 0x0F] | (BIT(data >> 6, 0) << 7));
+			output().set_value("trans_rate10", ls247_map[(m_trans_rate[0] >> 4) & 0x0F] | (BIT(data >> 5, 0) << 7));
+			output().set_value("trans_rate1", ls247_map[m_trans_rate[0] & 0x0F] | (BIT(data >> 4, 0) << 7));
+			output().set_value("pattern_number1000", (m_pattern_number[2] & 0x7F) | (BIT(data >> 3, 0) << 7));
+			output().set_value("pattern_number100", (m_pattern_number[1] & 0x7F) | (BIT(data >> 2, 0) << 7));
+			output().set_value("pattern_number10", ls247_map[(m_pattern_number[0] >> 4) & 0x0F] | (BIT(data >> 1, 0) << 7));
+			output().set_value("pattern_number1", ls247_map[m_pattern_number[0] & 0x0F] | (BIT(data >> 0, 0) << 7));
```
Tag lookups for outputs are also expensive.

Be consistent with hexadecimal case – you’re mixing uppercase and lowercase.  MAME developers tend to prefer lowercase.

felipesanches replied:
Not sure how to do it without these lookups. Is there an example of what you have in mind here?

felipesanches replied:
At this separate branch, I've refactored the code to use `output_finder`:
https://github.com/mamedev/mame/commit/aaeb409a399b7d19f225449048bb4c0a7bc222c5

But it seems that initiating output_finder with a list of tags is not supported yet.

I'm looking at `src/emu/devfind.h` as well as at `src/emu/output.h`. It seems that only devfind has support for initializing it with a list of tags. `output_manager::output_finder` does not seem to offer a similar method. It would be very useful though!
(correct me if I am wrong. I only quickly looked at these source files)

cuavas replied:
I’m still not convinced doing the outputs that way is a great idea.

I know this isn’t an issue when there’s only one piece of software on the platform, but having the names in code like this doesn’t scale out when you have more than one thing running on the system.  What the output actually connects to isn’t really the concern of the machine.

For a concrete example of drivers heavily dependent on outputs that work this way, see the Bally pinball drivers:
* The [by17.cpp](https://github.com/mamedev/mame/blob/mame0225/src/mame/drivers/by17.cpp) and [by35.cpp](https://github.com/mamedev/mame/blob/mame0225/src/mame/drivers/by35.cpp) drivers have output names that match the wiring diagrams.
* We have internal artwork for [Mata Hari](https://github.com/mamedev/mame/blob/mame0225/src/mame/layout/by17_matahari.lay), [PowerPlay](https://github.com/mamedev/mame/blob/mame0225/src/mame/layout/by17_pwerplay.lay) and [Playboy](https://github.com/mamedev/mame/blob/mame0225/src/mame/layout/by35_playboy.lay) that shows the playfield elements.
* We also have generic fallback artwork for showing the state of the outputs for the [17](https://github.com/mamedev/mame/blob/mame0225/src/mame/layout/by17.lay) and [35](https://github.com/mamedev/mame/blob/mame0225/src/mame/layout/by35.lay) boards.

Another example is the BFM Scorpion 2 system, where we have internal artwork for [Cops'n'Robbers Club](https://github.com/mamedev/mame/blob/mame0225/src/mame/layout/sc2copcl7.lay), [Premier Club Manager](https://github.com/mamedev/mame/blob/mame0225/src/mame/layout/sc2prem2.lay), [Along the Prom](https://github.com/mamedev/mame/blob/mame0225/src/mame/layout/sc2prom.lay), [Party Time](https://github.com/mamedev/mame/blob/mame0225/src/mame/layout/sc2ptytm1.lay), and some other machines, as well as generic artwork showing the lamp state in a grid, and the DMD and video output if present.

In practical terms, it’s more effort to modify or fix errors in the outputs if you attach semantics to them in the code.  At the very least, you need to go through a compile/link cycle to test a change.  If you’re just sending the raw values to outputs, you can just edit an external layout file to test a change, so the turnaround time for experimentation and testing is a lot faster.  It’s also more effort to keep a list of names in sync with the layouts than just a bunch of positions in the output matrix.

Until recently it wasn’t possible to extract bit fields from an output in the layout code, which would have caused issues for you used a single output for an 8-bit output matrix row.  However, that’s been addressed, and you can extract individual bits or bit fields with `statemask` for conditionally drawn elements, or `mask` for element position/colour animation.

Several internal layouts have been updated to use this functionality already, including:
* https://github.com/mamedev/mame/blob/00bfc6411ad8005edfbc347802d831793da7c177/src/mame/layout/unkeinv.lay#L31 (extracting a bit field from an input for item bounds animation)
* https://github.com/mamedev/mame/blob/00bfc6411ad8005edfbc347802d831793da7c177/src/mame/layout/esq2by16.lay#L35 (extracting individual bits from outputs for alphanumeric dot matrix displays)
* I also posted a simple demo of using `statemask` to allow the use of external graphics for 7-segment displays over at MAMEworld: https://www.mameworld.info/ubbthreads/showthreaded.php?Cat=&Number=388122&page=0&view=expanded&sb=5&o=&fpart=1&vc=1 (it also shows SVG in external artwork, but that’s a different kettle of fish)

But to answer the question about how to get an array of output finders with literal names, if you really need to, you can do it like this using C++ uniform initialisation:

```
class foo : ...
{
...
    output_finder<> m_outputs[10];
};
...
foo::foo(...)
...
    m_outputs{
        { *this, "first_name" },
        { *this, "second_name" },
        ...
        { *this, "tenth_name" } },
...
{
...
}
...
```

cuavas commented on src/mame/layout/dfs500.lay:
```diff
@@ -0,0 +1,286 @@
+	<!-- Snap Shot -->
+
+	<!-- Lighting -->
+	<element name="lighting_spot"  defstate="1"><image file="lighting_spot.png" state="1"/></element>
+	<element name="lighting_line"  defstate="1"><image file="lighting_line.png" state="1"/></element>
+	<element name="lighting_plane" defstate="1"><image file="lighting_plane.png" state="1"/></element>
+	<element name="lighting_width_wide"     defstate="1"><image file="lighting_width_wide.png" state="1"/></element>
+	<element name="lighting_width_medium"   defstate="1"><image file="lighting_width_medium.png" state="1"/></element>
+	<element name="lighting_width_narrow"   defstate="1"><image file="lighting_width_narrow.png" state="1"/></element>
+	<element name="lighting_intensity_high"    defstate="1"><image file="lighting_intensity_high.png" state="1"/></element>
+	<element name="lighting_intensity_medium"  defstate="1"><image file="lighting_intensity_medium.png" state="1"/></element>
+	<element name="lighting_intensity_low"     defstate="1"><image file="lighting_intensity_low.png" state="1"/></element>
```
You can’t require external files with internal artwork.  You’ll need to draw some basic buttons and indicators using the built-in components (mostly rect, disk and text).  There are plenty of examples you can learn from.

felipesanches replied:
done

MooglyGuy commented on src/mame/drivers/dfs500.cpp:
```diff
@@ -0,0 +1,1036 @@
+	return m_int_vector;
+}
+
+void dfs500_state::machine_start()
```
The majority of the functionality in this function belongs in `machine_reset` instead.

MooglyGuy commented on src/mame/drivers/dfs500.cpp:
```diff
@@ -0,0 +1,1036 @@
+	void cpanel_reg1_w(offs_t offset, uint8_t data);
+	void cpanel_reg2_w(offs_t offset, uint8_t data);
+	uint8_t cpanel_buzzer_r(offs_t offset);
+	u32 screen_update(screen_device &screen, bitmap_rgb32 &bitmap, rectangle const &cliprect);
```
Please do not mix shorthand (u8/u16/u32) and long-hand (uint8_t/uint16_t/uint32_t) types.

felipesanches replied:
which kind is preferred?

cuavas replied:
The rules for this are:
* Use the standard forms (`uint8_t` etc.) in self-contained things that we want to encourage use of outside MAME (e.g. strformat.{h,cpp} and vecstream.{h,cpp} in src/lib/util)
* Use the abbreviated forms (`u8` etc.) in src/emu
* For drivers and devices, choose a form and stick to it consistently throughout the source file (or files if there are more than one, e.g. the header and source for a device)

cuavas reviewed (changes requested):
This needs to be cleaned up.  You should also use more descriptive names for source files since not all of Sony’s Betacam and U-Matic products will be based on the same platform.

MooglyGuy commented on src/mame/drivers/dfs500.cpp:
```diff
@@ -0,0 +1,1036 @@
+uint8_t dfs500_state::pit_r(offs_t offset)
+{
+	// CXQ71054P (Programmable Timer / Counter)
+	return m_pit->read((offset>>1) & 3); // addressed by CPU's address bits AB2 and AB1
```
Please use consistent spacing with operators.

MooglyGuy commented on src/mame/drivers/dfs500.cpp:
```diff
@@ -0,0 +1,1036 @@
+	// FIXME! Add other signals.
+	if (m_TOC) value |= 0x02;
+	if (m_TOE) value |= 0x01;
+	value |= ((ioport("DSW_S3")->read() & 0x0f) << 8); // (Unknown)
```
Please use `required_ioport`. Do not use string lookup-based ioport access at runtime.

felipesanches replied:
done

MooglyGuy commented on src/mame/drivers/dfs500.cpp:
```diff
@@ -0,0 +1,1036 @@
+uint16_t dfs500_state::RB0_r(offs_t offset)
+{
+	return m_effectcpu_latch16;
+	m_TOC = false;
```
This line will never be executed as it occurs after the return, and will cause compile errors on some compilers. Please fix it.

felipesanches replied:
AWESOME! I think this is actually the reason for one of the bugs in the driver. Thanks for spotting this one. It is clearly incorrect, and I should have seen it, but... sometimes we don't see the simplest things ;-D

MooglyGuy commented on src/mame/drivers/dfs500.cpp:
```diff
@@ -0,0 +1,1036 @@
+
+void dfs500_state::cpanel_reg0_w(offs_t offset, uint8_t data)
+{
+	static const uint8_t ls247_map[16] =
```
I believe MAME already has output helpers for 7-segment displays. This function can be simplified.

felipesanches replied:
I'd be happy to learn more about it and adopt it. Do you know of an example in some other driver?

MooglyGuy replied:
To my surprise, we actually don't have output helpers for 7-segment displays hooked up to standard 74-series parts. Please ignore that bit of feedback, then. :)

MooglyGuy commented on src/mame/drivers/dfs500.cpp:
```diff
@@ -0,0 +1,1036 @@
+
+	// "Reg7_5":
+	//  Not sure yet of their actual use...
+	map(0x6B07a, 0x6B07b).mirror(0x90700).w("reg7_5", FUNC(generic_latch_16_device::write));
```
Please be consistent in your capitalization of hexadecimal constants. Here you have mixed both upper-case and lower-case in the same constant, even. Convention is typically to use lower-case hexadecimal constants.

MooglyGuy commented on src/mame/drivers/dfs500.cpp:
```diff
@@ -0,0 +1,1036 @@
+	PORT_BIT(0x02, IP_ACTIVE_LOW, IPT_KEYPAD) PORT_NAME("TRAIL/SHADOW TYPE") // SW10
+	PORT_BIT(0x01, IP_ACTIVE_LOW, IPT_KEYPAD) PORT_NAME("TRAIL") // SW9
+
+	PORT_START("RD12")
```
The entire unused mask can be specified on one line, specifying each individual bit as IPT_UNUSED is unnecessary.

felipesanches replied:
ok, cool. I did it this way just because I like the style of one line per bit.

MooglyGuy replied:
No problem, just wasn't sure if you knew that merged masks could be used. It's fine as it is.

MooglyGuy commented on src/mame/drivers/dfs500.cpp:
```diff
@@ -0,0 +1,1036 @@
+	IMAGE_PICTURE(config, m_input3);
+	IMAGE_PICTURE(config, m_input4);
+
+	SCREEN(config, m_screen, SCREEN_TYPE_RASTER);
```
If the service manual provides appropriate timing values, please use them via `set_raw`, as the combination of `set_refresh_hz`, `set_size`, and `set_visarea` are deprecated. If the raw timing values are not available, it's fine as it is.

felipesanches replied:
I haven't found that info yet. Maybe the service manual provides it, but I haven't seen it. It is a very long manual! :-)

If I ever find the info, I'll update the driver for sure

MooglyGuy commented on src/mame/drivers/dfs500.cpp:
```diff
@@ -0,0 +1,1036 @@
+ROM_END
+
+//   YEAR  NAME   PARENT/COMPAT MACHINE  INPUT    CLASS          INIT     COMPANY  FULLNAME                     FLAGS
+SYST(1994, dfs500,    0, 0,     dfs500, dfs500, dfs500_state, empty_init, "Sony", "DFS-500 DME Video Switcher", MACHINE_IS_SKELETON)
```
The driver has considerably more information in it than what is typically thought of as a skeleton driver. Consider adjusting to `MACHINE_NOT_WORKING`.

felipesanches replied:
ok, great! I'll do it

MooglyGuy reviewed (changes requested):
Overall, it looks pretty good. There are some minor issues that should be taken care of before merging, along with the concern mentioned by @Osso13 .

felipesanches commented:
> You should also use more descriptive names for source files since not all of Sony’s Betacam and U-Matic products will be based on the same platform.

@cuavas, For now that's the best I could come up with. And the betacam drivers are just skeletons. There's no hardware difference yet described on it. If it turns out that the hardware of each device is sufficiently different we might later split them into separate files with more specific filenames.

felipesanches commented:
I've addressed some of the feedback. The rest I'll have to deal with later since it would require a larger ammount of time and I need to focus on my actual work here ;-)

felipesanches commented:
thanks!

cuavas commented:
@ajrhacker you can’t merge stuff in this state.  We won’t be able to get to a state where we can have per-device outputs and have them saved properly if we allow new code to do this.  It’s not acceptable to continue to add legacy code and hope someone cleans it up later.

felipesanches commented:
What can I do here? I seem to have addressed all requests except one,  in which I was not sure how to proceed and asked for further guidance. @cuavas, could you please review my attempt to address the issue at aaeb409?

Should I open a new PR?

felipesanches commented:
> We won’t be able to get to a state where we can have per-device outputs and have them saved properly if we allow new code to do this. It’s not acceptable to continue to add legacy code and hope someone cleans it up later.

This sounds like I committed some major offense! Can you translate that into simpler english?

DavidHaywood commented:
@felipesanches I suggest just contributing to a project that appreciates your work instead, it's been clear for a long time that some people involved in MAME don't like you much, and this kind of nitpickery over minor things, and aggressively reverting submission that were fine by other devs + scolding the  just goes to highlight that.

I appreciate you trying, but it's kinda obvious at this point that you aren't wanted here.

cuavas commented:
> This sounds like I committed some major offense! Can you translate that into simpler english?

No, you haven’t committed a major offense in any sense.  I’ll explain what’s going on with this.  There’s a bit of history here.  Outputs in their current form are a bit ugly for a number of reasons, in large part due to the way they were hacked in.

Originally there were fixed arrays of outputs for “LEDs”, “lamps”, and maybe some other things, that were accessed by index.  They were just used for things like start button lamps, LED score displays, etc.  Mostly updated relatively infrequently and displayed on keyboard LEDs or using the artwork system when that arrived.  Drivers couldn’t control the names or total number of them, and they conceptually existed whether they were used or not.

Of course this ended up being inadequate, so the outputs were allowed to have arbitrary names.  However, a short-sighted decision was made to provide the same API where drivers just assumed outputs always existed and didn’t need to declare them upfront.  This has led to a number of ongoing problems:
* Writing to outputs became very slow.  Some things that were using outputs (e.g. the BFM LED matrix display) switched to using fake screens instead because the performance using outputs was terrible.
* Clients (e.g. MAMEhooker or LEDblinky) can’t check that the outputs they expect are present on start and need to deal with new outputs appearing at any time, because outputs are created when first written to.
* Output values cannot be saved/restored by the save state system, because the outputs don’t exist until written to.
* Outputs are global in a system, so multiple instances of the same device can cause issues.

Now some of the issues have been mitigated to some extent:
* Output finders allow outputs to be declared upfront.
* Reading or writing output values through output finders is very fast.
* Outputs that exist after `device_start` completes for all devices have their output values saved/restored in save states automatically.
* The BFM LED matrix display is happily using outputs again.

However, issues still remain:
* Outputs are still global, so they’re saved as a big blob for the system when we’d like to be able to associate the save data with individual devices, and the issues with multiple instances of the same device remain.
* We haven’t converted all drivers/devices over to using output finders, so we still have the issues with outputs being created after start.
* To make start-up deterministic, layout items have default state values applied after start, which can cause unexpected results in some cases.  (It can also potentially cause INP playback to fail or other issues depending on the artwork in use if a driver reads an output back and takes some action based on its value.)

Now because MAME is a big volunteer project, getting rid of any kind of legacy API is a lot of work and takes a long time.  No-one’s going and actively removing all the `set_value`/`get_value` calls from drivers at the moment, although some people are removing them in passing when they touch drivers for other reasons.  However, when the time comes, it’s going to be less work to remove `set_value`/`get_value` if we don’t add more of them.
